src/web/js/trove/d3-lib.js

```diff
@@ -397,70 +410,137 @@
-      image.onload = function () {
-          var opts = {
-              canvas: canvas.node(),
-              image: image
-          };
-          var a = document.createElement('a');
-          a.download = 'sample.png';
-          a.href = drawImage(opts).toDataURL('image/png');
-          a.click();
-
-          // the image's size was doubled everytime we click
-          // the button, so remove all data to prevent this
-          // to happen
-          svgData.remove();
-          canvas.remove();
+  function onSave(detached) {
+    var svgData = detached.append('div').style('display', 'none');
+    var imgsrc = getImageAsURL(detached);
+    var svg = detached.select('svg');
+    var canvas = detached.append('canvas')
+      .style('display', 'none')
+      .attr('width', svg.attr('width'))
+      .attr('height', svg.attr('height'));
+    svgData.html('<img src="' + imgsrc + '">');
+
+    var image = new Image;
+    image.src = imgsrc;
+    image.onload = function () {
+      var opts = {
+        canvas: canvas.node(),
+        image: image
       };
+      var a = document.createElement('a');
+      a.download = 'sample.png';
+      a.href = drawImage(opts).toDataURL('image/png');
+      a.click();
+
+      // somehow the image's size was doubled everytime we click
+      // the button, so remove all data to prevent this
+      // to happen
+      svgData.remove();
+      canvas.remove();
     };
   }
 
-  function callBigBang(detached, dimension, retVal, extra) {
+  function callBigBang(detached, restarter, resizer, windowOptions, dimension, retValFunc, extra) {
+    detached.select('.maing')
+      .append('text')
+      .attr('x', (dimension.marginLeft + dimension.width + dimension.marginRight) / 2)
+      .attr('y', dimension.height + dimension.marginTop + (8 * dimension.marginBottom / 11))
+      .html(libJS.htmlspecialchars(RUNTIME.getField(windowOptions, 'title')))
+      .style({
+        position: 'absolute',
+        'font-size': '8pt',
+        'text-anchor': 'middle'
+      });
+
     detached.selectAll('.overlay').style({
       fill: 'none',
       'pointer-events': 'all',
     });
 
-    RUNTIME.pauseStack(function(restarter) {
-      extra(restarter);
+    if (retValFunc === null) {
+      retValFunc = function(restarter) {
+        imageReturn(detached, restarter, function(x) { return x; });
+      }
+    }
+
+    if (RUNTIME.isPyretFalse(RUNTIME.getField(windowOptions, 'interact'))) {
+      RUNTIME.pauseStack(retValFunc);
+    }
+
+
+    var xscaler = libNum.scaler(
+      dimension.minWindowWidth,
+      dimension.maxWindowWidth,
+      0,
+      1
+    );
+
+    var yscaler = libNum.scaler(
+      dimension.minWindowHeight,
+      dimension.maxWindowHeight,
+      0,
+      1
+    );
+
+    var pauseStack;
+
+    if (RUNTIME.isNothing(restarter)) {
+      pauseStack = RUNTIME.pauseStack;
+    } else {
+      pauseStack = function(cb) { cb(restarter) };
+    }
+
+    pauseStack(function(restarter) {
```

Comment:
I don't understand this code at all -- what are you trying to do here?

pauseStack is for use when you are in a JS function being called while on a Pyret stack, but you need to do some JS processing that might be asynchronous, for example, so you need to pause Pyret until it resumes.  Inside the function being passed to pauseStack, you're _not_ on a Pyret stack any more, and you have no stack-safety guarantees; it's all managed by you.  You get back into the Pyret stack by calling the restarter.

safeCall is when you're in a JS function on a Pyret stack, and you want to call something that may eventually call into Pyret and so might bounce the stack.  The two functions passed to safeCall are the things to do before making the potentially-deep call back into Pyret, and then things to do afterward with the result of that call.

runThunk is for use when you're _not_ on a Pyret stack, and you want to call into a Pyret function that might bounce the stack.  Its signature is similar to safeCall, though the value passed into the second function is a Result, not a Pyret value, and needs some unwrapping to distinguish if things succeeded or failed.

There is no fourth function, for JS calls that might call into JS code that's asynchronous -- that's just life with JS.

Given those four scenarios, what is your code trying to do, and can you state clearly what its assumptions are about the kind of stack it's running in, and the kinds of functions it's calling?